File: solutions/typescript/queen-attack/1/queen-attack.ts
type Position = readonly [number, number]

type Positions = {
  white: Position
  black: Position
}

export class QueenAttack {
  public readonly black: Position
  public readonly white: Position

  constructor({ white = [7, 3], black = [0, 3] }: Partial<Positions> = {}) {
    // Check if positions are valid (within 0-7)
    if (!this.isValidPosition(white) || !this.isValidPosition(black)) {
      throw new Error('Queen must be placed on the board');
    }
    
    // Check if queens are on the same position
    if (white[0] === black[0] && white[1] === black[1]) {
      throw new Error('Queens cannot share the same space');
    }
    
    this.white = white;
    this.black = black;
  }

  private isValidPosition(position: Position): boolean {
    const [row, col] = position;
    return row >= 0 && row < 8 && col >= 0 && col < 8;
  }

  get canAttack(): boolean {
    const [whiteRow, whiteCol] = this.white;
    const [blackRow, blackCol] = this.black;
    
    // Same row
    if (whiteRow === blackRow) return true;
    
    // Same column
    if (whiteCol === blackCol) return true;
    
    // Same diagonal (difference in rows equals difference in columns)
    if (Math.abs(whiteRow - blackRow) === Math.abs(whiteCol - blackCol)) return true;
    
    return false;
  }

  toString(): string {
    const board: string[][] = [];
    
    // Initialize empty board
    for (let i = 0; i < 8; i++) {
      board.push(new Array(8).fill('_'));
    }
    
    // Place queens (row 0 is top row, row 7 is bottom row)
    const [whiteRow, whiteCol] = this.white;
    const [blackRow, blackCol] = this.black;
    
    board[whiteRow][whiteCol] = 'W';
    board[blackRow][blackCol] = 'B';
    
    // Convert to string representation
    return board.map(row => row.join(' ')).join('\n');
  }
}

File: solutions/r/queen-attack/1/queen-attack.R
create <- function(row, col) {
  if (row < 0 || row > 7 || col < 0 || col > 7) {
    stop("Position must be on the 8x8 chessboard (0-7).")
  }
  c(row = row, col = col)
}

can_attack <- function(queen1, queen2) {
  if (all(queen1 == queen2)) {
    stop("Queens cannot occupy the same position.")
  }
  
  # Same row or same column
  if (queen1["row"] == queen2["row"] || queen1["col"] == queen2["col"]) {
    return(TRUE)
  }
  
  # Same diagonal (absolute difference in rows equals absolute difference in columns)
  if (abs(queen1["row"] - queen2["row"]) == abs(queen1["col"] - queen2["col"])) {
    return(TRUE)
  }
  
  return(FALSE)
}